### PR TITLE
ValidationItem Initializer Fix

### LINKED ISSUE #2758: Error when edit button clicked in datagrid row with a complex object property
https://support.blazorise.com/issues/5/Error-when-edit-button-clicked-in-datagrid-row-with-a-complex-object-property

### PR DESCRIPTION
**Pull request by David-Moreira**
Fixes https://support.blazorise.com/issues/5/Error-when-edit-button-clicked-in-datagrid-row-with-a-complex-object-property
Closes #2758

As we've talked about, the problem was on the internal ValidationItem object being initialized from Datagrid's TItem as a base, acting as a dummy for validation purposes.

Current solution Initializes every public instance, except valuetypes, string and lists.
Added unit tests to support the implementation.
Benchmark is okay, but it is of course heavier then the previous simple Activator.
image

Code used to reproduce original problem:
```
@page "/"
<DataGrid TItem="MarketDataModel"
          Data="@BpsDataWithMarketData" Class="responsive-table table" ShowCaptions="true"
          Filterable="true" PageSize="@BpsDataWithMarketData.Count()" Striped="true"
          TotalItems="@BpsDataWithMarketData.Count()" Narrow="true" Editable="true" EditMode="DataGridEditMode.Inline"
          UseInternalEditing="true">
    <DataGridNumericColumn TItem="MarketDataModel" Field="bpsData.UnderlyingSecurityDetails.BuySellLevel.SellQuantity" Caption="Sell #" Sortable="true" Editable="true" Filterable="false">
    </DataGridNumericColumn>
    <DataGridCommandColumn TItem="MarketDataModel" ShowCaption="true" Caption="captiontodisplay" DeleteCommandAllowed="false" FilterCellStyle="position:sticky; top: 110px; line-height:1em;padding:1;" NewCommandAllowed="false" EditCommandAllowed="true">

        <EditCommandTemplate>
            <Blazorise.Button Color="Color.Primary" Clicked="@context.Clicked" Margin="Margin.Is0">Edit</Blazorise.Button>
        </EditCommandTemplate>
        <SaveCommandTemplate>
            <Blazorise.Button Color="Color.Primary" Clicked="@context.Clicked" Margin="Margin.Is0">Save</Blazorise.Button>
        </SaveCommandTemplate>
        <CancelCommandTemplate>
            <Blazorise.Button Color="Color.Primary" Clicked="@context.Clicked" Margin="Margin.Is0">Cancel</Blazorise.Button>
        </CancelCommandTemplate>
    </DataGridCommandColumn>
</DataGrid>
@code {
    private List<MarketDataModel> BpsDataWithMarketData = new()
    {
        new()
        {
            bpsData = new()
            {
                UnderlyingSecurityDetails = new()
                {
                    BuySellLevel = new()
                    {
                        SellQuantity = 1
                    }
                }
            }
        }
    };
    public class MarketDataModel
    {
        public BpsDataModel bpsData { get; set; } 
    }
    public class BpsDataModel
    {
        public SecurityDetails UnderlyingSecurityDetails { get; set; }
    }

    public class SecurityDetails
    {
        public BuySellLevel BuySellLevel { get; set; }
    }

    public class BuySellLevel
    {
        public int SellQuantity { get; set; }
    }
}
```

### CODE REVIEW

File: Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace Blazorise.DataGrid.Utils
+{
+    public static class Reflection
```
Comment (stsrki) [inline Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs:7]: Name it NestedActivator, ObjectInitializer, or something similar?

File: Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        /// <summary>
+        /// Inits a complex object and inner public instance properties.
+        /// Ignores lists, valuetypes and strings
+        /// </summary>
+        /// <typeparam name="TItem"></typeparam>
+        /// <returns></returns>
+        public static TItem InitObject<TItem>()
```
Comment (stsrki) [inline Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs:15]: Name it CreateInstance so it is along the Activator naming convention.

File: Tests/Blazorise.Benchmark/Program.cs [hunk 3]
```diff
@@ -4,29 +4,30 @@
...
+
             [Benchmark]
             public bool SequenceEqual() => simpleList1.SequenceEqual( simpleList2 );
 
             [Benchmark]
             public bool ArraySequenceEqual() => simpleList1.AreEqual( simpleList2 );
         }
     }
-}
+}
```
Comment (stsrki) [inline Tests/Blazorise.Benchmark/Program.cs:33]: I get a lot of _noise_ in PR because of this line ending. Did you use some kind of auto-format?

Comment (David-Moreira) [reply Tests/Blazorise.Benchmark/Program.cs:33]: I ran a tool, would you prefer that I revert?

File: Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs (hunk 1 above)
Comment (David-Moreira) [reply Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs:7]: I like ObjectInitializer or even ObjectActivator.

File: Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs (hunk 2 above)
Comment (David-Moreira) [reply Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs:15]: Sure,

File: Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs (hunk 1 above)
Comment (stsrki) [reply Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs:7]: RecursiveActivator?

File: Tests/Blazorise.Benchmark/Program.cs (hunk 3 above)
Comment (stsrki) [reply Tests/Blazorise.Benchmark/Program.cs:33]: Yeah, its easier to review pr when there are less changes

File: Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs (hunk 1 above)
Comment (David-Moreira) [reply Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs:7]: Or RecursiveObjectActivator? haha naming is hard! I kinda like the Object prefix, cuz you enunciate that you are activating an object.

Comment (stsrki) [reply Source/Extensions/Blazorise.DataGrid/Utils/Reflection.cs:7]: Haha yeah, its always hard to name things. RecursiveObjectActivator is good

**Comment by stsrki** [comment]
Looking good now. Performance doesn't matter too much since this is called only once on the Edit command. So as long as it creates objects properly all is good.

Nice work!

**Pull request merged**